pennylane/templates/subroutines/trotter.py

```diff
@@ -0,0 +1,216 @@
+        check_hermitian (bool): A flag to enable the validation check to ensure this is a valid unitary operator.
+
+    Raises:
+        TypeError: The 'hamiltonian' is not of type ~.Hamiltonian, or ~.Sum.
+        ValueError: One or more of the terms in 'hamiltonian' are not Hermitian. 
+        ValueError: The 'order' is not one or a positive even integer.
+
+    **Example**
+
+    .. code-block:: python3
+
+        coeffs = [0.25, 0.75]
+        ops = [qml.PauliX(0), qml.PauliZ(0)]
+        H = qml.dot(coeffs, ops)
+
+        dev = qml.device("default.qubit", wires=2)
+        @qml.qnode(dev)
+        def my_circ():
+            # Prepare some state
+            qml.Hadamard(0)
+
+            # Evolve according to H
+            qml.TrotterProduct(H, time=2.4, order=2)
+
+            # Measure some quantity
+            return qml.state()
+
+    >>> my_circ()
+    [-0.13259524+0.59790098j  0.        +0.j         -0.13259524-0.77932754j  0.        +0.j        ]
+
+    .. details::
+        :title: Usage Details
+
+        We can also compute the gradient with respect to the coefficients of the hamiltonian and the
+        evolution time: 
+
+        .. code-block:: python3
+
+            @qml.qnode(dev)
+            def my_circ(c1, c2, time): 
+                # Prepare H: 
+                H = qml.dot([c1, c2], [qml.PauliX(0), qml.PauliZ(1)])
+                    
+                # Prepare some state
+                qml.Hadamard(0)
+                
+                # Evolve according to H
+                qml.TrotterProduct(H, time, order=2)
+                
+                # Measure some quantity
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        >>> args = qnp.array([1.23, 4.5, 0.1])
+        >>> qml.grad(my_circ)(*tuple(args))
+        (tensor(0.00961064, requires_grad=True), tensor(-0.12338274, requires_grad=True), tensor(-5.43401259, requires_grad=True))
+    """
+
+    def __init__(self, hamiltonian, time, n=1, order=1, check_hermitian=True, id=None):
+        r"""Initialize the TrotterProduct class"""
+
+        if order <= 0 or order != 1 and order % 2 != 0:
+            raise ValueError(
+                f"The order of a TrotterProduct must be 1 or a positive even integer, got {order}."
+            )
+
+        if isinstance(hamiltonian, qml.Hamiltonian):
+            coeffs, ops = hamiltonian.terms()
+            hamiltonian = qml.dot(coeffs, ops)
+
+        if not isinstance(hamiltonian, Sum):
+            raise TypeError(
+                f"The given operator must be a PennyLane ~.Hamiltonian or ~.Sum got {hamiltonian}"
+            )
+
+        if check_hermitian:
+            for op in hamiltonian.operands:
+                if not op.is_hermitian:
+                    raise ValueError(
+                        "One or more of the terms in the Hamiltonian may not be hermitian"
+                    )
+
+        self._hyperparameters = {"num_steps": n, "order": order, "base": hamiltonian}
```

Comment:
I'm confused about the "base" name

Reply:
I just use that to denote the input hamiltonian, it was shorter than typing "hamiltonian" every time 😄